### DISCUSSION TITLE
Panel migration

### DISCUSSION THREAD

**Discussion by spencerwahl** [fgpv-vpgf/fgpv-vpgf#4109] [Team Posts]
Current situation
-----
- `stateManager` coordinates all panel opening and closing
- `stateManager` also manages and stores the panel state (this part fits the name)
- `displayManager` handles coordinating what data goes where (this shouldn't need any altering)
- Opening and closing panels affects other panels depending on relation; i.e. closing legend closes the settings panel
- Panels are created at app startup and directives take care of the display part

Goal
----
- Maintain panel interactions; any panel that automatically opened/closed another should continue to do so
- The `appbar` should continue to have its label updated (Layers, Details, etc.) depending on what panel is open
- Be able to open, close and toggle panels
- Code maintainability (avoiding a web of panels calling eachother and having to know the state of the rest of the app)

What we have
----
- Panel API can close panels automatically if they get overlapped, this is good for interactions such as opening settings with the table open, etc.
- Panels expose two observables that fire on the panel opening or closing
- We can put the angular directive tag in the panel and it will work as before

Issues
----
- Certain panel interactions close other panels without overlapping, this requires something to coordinate panel interactions
- `stateManager` doesn't know about the new panels close buttons/automatic overlap close

Possible ways forward
----
### 1. We keep state manager doing what its currently doing

- `stateManager` manages state
- `stateManager` manages panel interactions
- Panels react to what `stateManager` tells them to do

Pros:
- A lot is written for us already
- things like the appbar only need to grab panel state from one place

Cons:
- Need to change panel state tracking, panels can close using their own close button (if existing). Either need to subscribe to the closing (and possibly opening) observable on each panel, or figure out another way
- Still need to change how `stateManager` views panels, there are no longer parent/child panels. (Main panel vs legend panel)
- `stateManager` stays confusing

### 2. We maintain state within each panel

- Panel stores open/closed state
- Service for the component listens on observable telling which panel should end in what state, panel opens/closes if needed
- `stateManager` handles interactions between panels, 

Pros:
- Makes sense for the panel to know whether its opened or closed
- Separate the responsibilities of `stateManager`, it would now be a panel coordinator (A name change will probably be desired)

Cons:
- Decentralized state management (at least I think thats a con)
- Related; if we want to put panel state on the bookmark/in whatever state storage we have in the future, will need extra dev
- Things like the appbar need to get panel state, maybe we have a centralized panel registry that exposes things like the currently open panels, etc.

### 3. Panels handle panel interactions

- example: `stateManager` says legend should close -> legend service tells settings/metadata to close if they're open, or settings listens on legend close and closes
- panels maintain their own state, `stateManager` has no idea what its doing (its probably removed in this case)

Pros:
- If you really dislike `stateManager` its gone
- Components are responsible for their own interactions

Cons:
- Components are responsible for their own interactions
- Old code is not useful
- panels have to know other panels state (`geoSearch` opens and legend closes, needs to know to open legend when it closes)

**Comment by milespetrov** [comment]
Repost from separate thread.

The following is a general outline of modifications to the panel API as well as coding flow changes, should we choose to move away from state manager, that would support more complex panel state requirements. 

## New panel api option - **reopenAfterOverlay: boolean**

When `true`, a panel that was closed due to an overlaying panel is reopened as soon as space is available. This option is **false** by default. This is similar to state managers **history** functionality. 

Example: geosearch closes the details panel by overlaying it, but if the details panel has `reopenAfterOverlay = true` it would reopen as soon as geosearch is closed.

## Opacity documentation

geosearch closes the `main` panel but does not display anything until results are available. This scenario is currently possible by setting the opacity of the geosearch result panel to `0`. This way the panel api will still treat the panel as **open** (and proceeds to apply overlay rules, closing the `main` panel`) while keeping it hidden from the user. We'd just need to document this as an "advanced topic" for certain use cases.

## Panels with linked state

When the legend is closed, for example, the settings panel closes as well. This functionality would now belong with the settings panel code. The settings panel would watch the legend panels `closing` observable and close itself as needed.

**Comment by james-rae** [comment]
Random thought on the appbar aspect (caveat - I know nothing about panel api):

> The appbar should continue to have its label updated

> Things like the appbar need to get panel state

Almost feels like "I influence the appbar" is a panel property.  Some panels won't (e.g. panel holding Area Of Interest).  Would a push-approach be more effective?

E.g.

```text
Class Panel {
  open() {
      if (this.iInfluenceTheAppbar) {
         panelSheriff.appbar.setNewActivePanel(this.appbarName);
      }
      ...
  }

  close() {
      if (this.iInfluenceTheAppbar) {
         // clear appbar or pop a stack of appbar names
         // or ask sheriff to find any other open magic panels
      }
  }
```

**Comment by ShrutiVellanki** [comment]
To elaborate on how  #2 might work, I'll just repost what I wrote yesterday :)

### PanelCoordinator

`PanelCoordinator` deals with rules of open/close interactions between panels. 

1. `PanelCoordinator` subscribes to `open`/`close` observables on each panel
2. It would coordinate interactions between panels by reacting to opening and closing panels in one of two ways: 
    #### 1. Standard Interactions:
    The standard underlay/offscreen rules for most panels.
    
    #### 2. Special Interactions
    Coordinates using special rules by storing recent history. For example  it would contain the code for: ``"if geosearch is opened, the legend would close, and once its closed, the legend reopens"``
    
----

### Panel State Manager 

Each panel would have its own `PanelStateManager`. A version of this is implemented for `enhancedTables` [here](https://github.com/fgpv-vpgf/plugins/blob/master/enhancedTable/panel-state-manager.ts). 

- Each `close` call for a panel wil "save" its state to its `PanelStateManager`
- Each `open` call for a panel will "load" state from its `PanelStateManager`

#### External Modification of State
Since `PanelStateManager` would be available on each `Panel` object, you can access it easily as long as you have the `id` of the Panel.
-  **Pro:** Modifying states externally (for example for a plugins author) would be possible. 
- A modification of state would trigger a reload of the display for all open panels

**Comment by spencerwahl** [comment]
Notes from meeting on March 22, 2019:

- Centralized panel coordination looks like easier debugging but much more complicated logic
- Decentralized panel coordination looks like easier logic but a longer chain to follow for debugging, easier to allow plugins to fit their way in

The team has chosen Decentralized panel coordination (aka something that looks like option 3).

- This will incorporate the functionality Miles posted above; `reopenAfterOverlay` and special geosearch fun.
- we will expose a function that allows manipulation of the appbar title

`PanelRegistry` class:
- moves panel array from map instance to this part of api
- handles anything that needs to deal with all of the panels, such as....

`reopenAfterOverlay`:
- maintain a stack that panels are pushed onto when overlayed AND they have `reopenAfterOverlay` set to true
- When panel configuration changes, run through stack and see if anything can be reopened
- if panel can be opened; pop it and continue through the stack

**Comment by spencerwahl** [comment]
List of current panel interactions that were written up;

## Legend

### open

- closes any other main panel

### close

- closes metadata/settings, closes main panel

## Details

### open

- closes metadata, settings (doesn't currently but should), closes geosearch & covers anything main panel

### close

- close button -> reopens legend if needed
- appbar toggle -> closes main panel fully

## Settings/metadata

### open

- closes table

### close

- closes side panel (no history between settings & metadata)

## Import

### open

- technically covers legend

### close

- reopens legend

## Table

### open

- closes settings/metadata

### close

- nothing

## GeoSearch

### open

- closes everything except table

### close

- reopens last main panel if it was open
